cmsl1t/config.py

```diff
@@ -250,6 +253,47 @@ def dump(self, out_filename):
         with open(out_filename, "w") as out_file:
             out_file.write(yaml.dump(config))
 
+    def _reduce_scopes(self):
+        '''
+            Reduces config scopes for analyzers and producers
+        '''
+        cfg = self.config
+        analyzers = self.get('analysis', 'analyzers')
+        analyzers = [self.reduce_scope_for_analyzer(a) for a in analyzers]
+        cfg['analysis']['analyzers'] = analyzers
+
+        producers = self.get('analysis', 'producers')
+        producers = [self.reduce_scope_for_producer(p) for p in producers]
+        cfg['analysis']['producers'] = producers
+
+    def reduce_scope_for_analyzer(self, analyzer_name):
+        analyzer = self.get('analysis', 'analyzers')[analyzer_name]
+        global_settings = dict(
+            output_folder=self.get('output', 'folder'),
+            plots_folder=self.get('output', 'plots_folder'),
+            puBins=self.try_get('analysis', 'pu_bins', [0, 999]),
+            thresholds=self.try_get('analysis', 'thresholds', None),
+            file_format=self.try_get('output', 'plot_format', 'pdf'),
+            triggerName=self.get('input', 'trigger')['name'],
+            lumiJson=self.try_get('input', 'lumi_json', ''),
+            # TODO: this one should be removed after refactoring jetMetAnalyzer
+            load_trees=self.try_get('analysis', 'load_trees'),
+            # TODO: do better for legacy analyzer
+            input_files=self.get('input', 'files'),
+        )
+        reduced_scope = {'name': analyzer_name}
+        reduced_scope.update(global_settings)
+        reduced_scope.update(analyzer)
```

Comment:
The order of precedence here could lead to some confusing situations.  Eg. if you define a key-value pair in the analyzer's arguments with the key == `name` that will over-ride the name given to the actual analyzer using the normal approach.  Similarlry, since `input/files` is now passed to the analyzer directly, in principle one could over-ride the global value for this analyzer by redefining it in the analyzer's config.  Not sure what would happen then, but would probably not be what was expected by the user.

Reply:
Indeed, you can get into some weird corners.
However, in many areas, this is what you expect: Global defaults  & specialisation in the modules (e.g. hiera). It really depends on what we are trying to do.

With the final goal to vectorise the event-loop, we should ban overwriting the input_files (only global value is valid)

Reply:
I think it just needs a degree of validation / sanitization.  Certain keys (eg. `name` and `input_files`) would be banned from the users configuration of each analyzer.  If provided, at this point in the code we detect and raise an exception.

Reply:
OK, done